### PR TITLE
feat: Filter projects by quarter

### LINKED ISSUE #1648: Projects view: Filter by Q
**Sidebar interaction**

Image

https://www.figma.com/file/e0IFJjplUdYdQJ2NtLwvMf/DAO-v2-(Guane)?type=design&node-id=15335-5874&mode=dev

**Projects view, when filters are interacted with**

- [ ] big purple module with links at the top is to become collapsible
- [ ] data visualized is that of the selected Quarter
- [ ] when filtering "Grant Category" + certain quarter, visualization changes

Image

Image

See endpoing #1599

### PR DESCRIPTION
**Pull request by ncomerci**
**Screenshots**
image
image

**Related Issue**
This closes #1648 

- [ ]  **Functionality validation**
    - [ ]  Feature working as described in the related issue
    - [ ]  New environment variables have been set
    - [ ]  New API services methods have tests
    - [ ]  Code errors are being logged in the code health service
- [ ]  **Product validation**
    - [ ]  UI changes look the same as in Figma
    - [ ]  Usage events are being sent to the product analytics service
    - [ ]  New common components have been documented in Storybook
    - [ ]  Preview link if it’s a customer facing issue
- [ ] **Documentation**
    - [ ] README updated
    - [ ] New API changes are documented
- [ ] **Accessibility**
    - [ ] UI changes are accessible

### CODE REVIEW

File: src/pages/projects.tsx [hunk 1]
```diff
@@ -80,15 +81,26 @@ function toProjectTypeFilter(category?: string | null): ProjectTypeFilter | unde
...
   const t = useFormatMessage()
   const params = useURLSearchParams()
   const type = toProjectTypeFilter(params.get('type'))
   const status = toProjectStatus(params.get('status'))
   const subtype = toGrantSubtype(params.get('subtype'), () => undefined)
+  const year = toDateFilter(params.get('year'))
+  const quarter = toDateFilter(params.get('quarter'))
 
-  const { projects, isLoadingProjects } = useProjects()
+  const { projects, isLoadingProjects } = useProjects(year && quarter ? { year, quarter } : undefined)
```
Comment (andyesp) [inline src/pages/projects.tsx:103]: If year and quarter could be undefined and checks are added in the hook I'd just write here `useProjects({ year, quarter })`

File: src/components/Projects/Current/StatsAllProjects.tsx [hunk 2]
```diff
@@ -25,11 +26,16 @@ export default function StatsAllProjects({ projects }: Props) {
     return finishAt.isAfter(Time()) && finishAt.isBefore(Time().add(1, 'week'))
   })
 
-  const currentQuarter = Time().quarter()
+  const { year: yearParam, quarter: quarterParam } = useYearAndQuarter()
```
Comment (1emu) [inline src/components/Projects/Current/StatsAllProjects.tsx:29]: If `useYearAndQuarter` already checks for

```typescript
 const year = validateYear(yearParam) ?? Time().year()
  const quarter = validateQuarter(quarterParam) ?? Time().quarter()
  ``` 
  
  then you can avoid this part right?
  
  ```typescript
  const isYearAndQuarterValid = yearParam && quarterParam

  const currentYear = isYearAndQuarterValid ? yearParam : Time().year()
  const currentQuarter = isYearAndQuarterValid ? quarterParam : Time().quarter()
  ```

File: src/components/Projects/Current/StatsBiddingAndTendering.tsx [hunk 3]
```diff
@@ -22,11 +23,17 @@ export default function StatsBiddingAndTendering({ projects }: Props) {
   const t = useFormatMessage()
   const formatFundingValue = (value: number) => intl.formatNumber(value, CURRENCY_FORMAT_OPTIONS)
 
-  const currentQuarter = Time().quarter()
+  const { year: yearParam, quarter: quarterParam } = useYearAndQuarter()
+  const isYearAndQuarterValid = yearParam && quarterParam
```
Comment (1emu) [inline src/components/Projects/Current/StatsBiddingAndTendering.tsx:27]: Same here :)

**Review by 1emu — APPROVED** [review]

**Review by 1emu — APPROVED** [review]

File: src/hooks/useBudgetByCategory.ts [hunk 4]
```diff
@@ -24,14 +26,37 @@ function convertPercentageToInt(percentage: number): number {
...
+export default function useBudgetByCategory(category: SubtypeOptions, year?: number | null, quarter?: number | null) {
+  const { data: allBudgets } = useQuery({
+    queryKey: ['all-budgets'],
+    queryFn: () => Governance.get().getAllBudgets(),
     staleTime: DEFAULT_QUERY_STALE_TIME,
   })
   const [budget, setBudget] = useState(INITIAL_BUDGET)
 
+  const currentYear = year ?? Time().year()
+  const currentQuarter = quarter ?? Time().quarter()
```
Comment (andyesp) [inline src/hooks/useBudgetByCategory.ts:46]: Maybe `selectedYear`, `selectedQuarter`?

File: src/hooks/useProjects.ts [hunk 5]
```diff
@@ -1,13 +1,26 @@
...
+
+export default function useProjects(dateFilter?: DateFilter) {
+  const { year, quarter } = dateFilter || {}
   const { data: projects, isLoading: isLoadingProjects } = useQuery({
-    queryKey: ['projects'],
-    queryFn: async () => Governance.get().getProjects(),
+    queryKey: ['projects', year, quarter],
+    queryFn: async () => {
+      const { startDate, endDate } =
+        quarter && year ? getQuarterDates(quarter, year) : { startDate: undefined, endDate: undefined }
```
Comment (andyesp) [inline src/hooks/useProjects.ts:19]: I think you could write this ```const { startDate, endDate } = quarter && year ? getQuarterDates(quarter, year) : {}```

File: src/hooks/useYearAndQuarter.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import { useMemo } from 'react'
+
+import { useLocation } from '@reach/router'
+
+import { validateQuarter, validateYear } from '../helpers'
+
+function useYearAndQuarter() {
```
Comment (andyesp) [inline src/hooks/useYearAndQuarter.ts:7]: `useYearAndQuarterParams`?

File: src/components/Projects/Current/StatsAllProjects.tsx (hunk 2 above)
Comment (andyesp) [reply src/components/Projects/Current/StatsAllProjects.tsx:29]: You could also return an `isValid` property from the hook as you are also doing that check in another component.

File: src/components/Projects/Current/StatsAllProjects.tsx [hunk 7]
```diff
@@ -25,11 +26,16 @@ export default function StatsAllProjects({ projects }: Props) {
...
+  const isYearAndQuarterValid = yearParam && quarterParam
+
+  const currentYear = isYearAndQuarterValid ? yearParam : Time().year()
+  const currentQuarter = isYearAndQuarterValid ? quarterParam : Time().quarter()
```
Comment (andyesp) [inline src/components/Projects/Current/StatsAllProjects.tsx:33]: selectedYear, selectedQuarter?

File: src/components/Search/QuarterFilter.tsx [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+  return years
+}
+
+function QuarterFilter() {
+  const t = useFormatMessage()
+  const [selectedYear, setSelectedYear] = useState<number | undefined>()
+  const { year: yearParam, quarter: quarterParam } = useYearAndQuarter()
+  const location = useLocation()
+  const params = useMemo(() => new URLSearchParams(location.search), [location.search])
```
Comment (andyesp) [inline src/components/Search/QuarterFilter.tsx:35]: You can use here `useURLSearchParams`

File: src/components/Projects/Current/BudgetBanner.tsx [hunk 9]
```diff
@@ -41,11 +42,13 @@ export const getNewGrantIcon = (type: string, variant?: CategoryIconVariant, siz
 export default function BudgetBanner({ category, status, initiativesCount }: Props) {
   const t = useFormatMessage()
   const intl = useIntl()
+  const { year, quarter } = useYearAndQuarterParams()
+  const isYearAndQuarterValid = year && quarter
```
Comment (andyesp) [inline src/components/Projects/Current/BudgetBanner.tsx:46]: Why doesn't the hook return a "areValidParams" variable?

File: src/hooks/useYearAndQuarterParams.ts [hunk 10]
```diff
@@ -0,0 +1,21 @@
+import { useMemo } from 'react'
+
+import { useLocation } from '@reach/router'
+
+import { validateQuarter, validateYear } from '../helpers'
+
+function useYearAndQuarterParams() {
+  const location = useLocation()
+  const params = useMemo(() => new URLSearchParams(location.search), [location.search])
```
Comment (andyesp) [inline src/hooks/useYearAndQuarterParams.ts:9]: `useURLSearchParams`?

**Review by andyesp — APPROVED** [review]

**Pull request merged**